consensus/core/src/dag_state.rs

```diff
@@ -234,7 +234,25 @@ impl DagState {
         }
     }
 
+    // Write commits to store. Commits should be provided in commit order, meaning
+    // the last element in commits is the new last_commit.
+    pub(crate) fn write_commits(
+        &mut self,
+        commits: Vec<Commit>,
+        committed_blocks: Vec<VerifiedBlock>,
+    ) {
+        assert!(!commits.is_empty());
+        let last_commit = commits.last().unwrap().clone();
+        self.store
+            .write(committed_blocks, commits)
+            .expect("Writing commits to store should not fail");
+        self.set_last_commit(last_commit);
+    }
+
     pub(crate) fn set_last_commit(&mut self, commit: Commit) {
+        if let Some(last_commit) = &self.last_commit {
+            assert!(commit.index >= last_commit.index);
```

Comment:
In which case we expect to be `commit.index == last_commit.index` ? Given the recovery path and the common case I believe we should always move forward , right?

Reply:
It is very subtle, but the way it works currently is that the linearizer batch committed subdags to write to store but it will update the last commit in dag state on each collected subdag so that the next collected subdag can use that as a reference to ensure no duplicate blocks are collected. So once the batch of subdags are eventually written to store the last commit in dagstate will be equal to the last one already set by linearizer during collection.

Can follow up in the next PR if there are concerns around this.